### PR TITLE
[67642] Avoid error when attachment is not uploaded yet

### PR DESCRIPTION
**Pull request by cbliard**
# Ticket

https://community.openproject.org/wp/67642

# What are you trying to accomplish?

2 things:
- when an error occurs in `Attachments::ExtractFulltextJob`, see it in AppSignal (do not swallow the error)
- prevent the error "NoMethodError: undefined method 'body' for nil:NilClass" from happening in `Attachments::ExtractFulltextJob`

# What approach did you choose and why?

`Attachments::ExtractFulltextJob` is run twice:
- one time on creation
- one time when the direct upload is completed

When it runs on creation, and the attachments are stored on S3, the attachment is in 'prepared' status and is not complete yet.

Due to a bug in carrierwave which is fixed since June 2023 (see https://github.com/carrierwaveuploader/carrierwave/issues/2524), the `#local_file` method raises the error "NoMethodError: undefined method 'body' for nil:NilClass". There is a separate issue for upgrading this dependency one day: https://community.openproject.org/wp/67626.

The fix is to call `#local_file` only if the attachment is readable.

Additionnally:
- error handling has been updated to raise the error instead of swallowing it silently, so well have proper reporting in AppSignal next time.
- when a custom S3 endpoint is used (for local testing with minio for instance), this custom endpoint is added to the CSP.

# Merge checklist

- [x] Added/updated tests
- [ ] Added/updated documentation in Lookbook (patterns, previews, etc)
- [ ] Tested major browsers (Chrome, Firefox, Edge, ...)

### CODE REVIEW

File: lib_static/open_project/configuration/helpers.rb [hunk 1]
```diff
@@ -131,13 +131,17 @@ def use_iam_profile?
       end
 
       def fog_credentials
-        (Hash(self["fog"])["credentials"] || {}).map { |key, value| [key.to_sym, value] }.to_h
+        (Hash(self["fog"])["credentials"] || {}).transform_keys(&:to_sym).to_h
```
Comment (NobodysNightmare) [inline lib_static/open_project/configuration/helpers.rb]: :green_circle: What about

```suggestion
        (Hash(self["fog"])["credentials"] || {}).symbolize_keys
```

File: spec/workers/attachments/extract_fulltext_job_spec.rb [hunk 2]
```diff
@@ -102,28 +113,40 @@
...
 
-    it "logs the error" do
-      extracted_attributes
-      expect(logger).to have_received(:error).with(/boom-internal-error/)
+    it "raises the exception to update Job internal state and logs it" do
+      expect { perform_enqueued_jobs }.to raise_error(exception_message)
     end
 
-    include_examples "only file name indexed"
+    context "even with an exception being raised" do
```
Comment (NobodysNightmare) [inline spec/workers/attachments/extract_fulltext_job_spec.rb]: Looking at the `before`, is this a typo?

```suggestion
    context "even without an exception being raised" do
```

**Review by NobodysNightmare — APPROVED** [review]
Only stylistic comments from my side. The changes look otherwise good at achieving the goals of:

* making unexpected errors louder
* avoiding too loud handling of an expected error case

File: spec/workers/attachments/extract_fulltext_job_spec.rb (hunk 2 above)
Comment (cbliard) [reply spec/workers/attachments/extract_fulltext_job_spec.rb]: No it's not a typo: the job is executed and raises an exception. It's rescued in the test because we don't want the test to end here.
And even if the job has raised an exception, a tsv has been added from the attachment filename, and the error was logged.

File: lib_static/open_project/configuration/helpers.rb (hunk 1 above)
Comment (cbliard) [reply lib_static/open_project/configuration/helpers.rb]: Yep, even better 👍

File: spec/workers/attachments/extract_fulltext_job_spec.rb (hunk 2 above)
Comment (NobodysNightmare) [reply spec/workers/attachments/extract_fulltext_job_spec.rb]: I think what I find surprising about reading this context is that the raising of the exception is not related to the context. The parent context "with exception during extraction" is the one that introduces the raising of the exception, so it's not quite clear to me what behavioural change this context here is adding then. I guess none, except for "test implementation"-specifics.

My usual assumption about a context is, that it changes the test setup/premise, which it didn't here.

File: lib_static/open_project/configuration/helpers.rb (hunk 1 above)
Comment (cbliard) [reply lib_static/open_project/configuration/helpers.rb]: And it also fixes an issue that I do not quite understand:

```
>> Hash(self["fog"])["credentials"] || {}
=> {"provider" => "AWS"}
>> (Hash(self["fog"])["credentials"] || {}).transform_keys(&:to_sym)
=> {"provider" => "AWS"}
>> {"provider" => "AWS"}.transform_keys(&:to_sym)
=> {provider: "AWS"}
```

I would have expected `(Hash(self["fog"])["credentials"] || {}).transform_keys(&:to_sym)` to return `{provider: "AWS"}`, not `{"provider" => "AWS"}`.

Comment (NobodysNightmare) [reply lib_static/open_project/configuration/helpers.rb]: Can you do `Hash(self["fog"])["credentials"].class`?

I bet it's not a hash, but something that _mostly_ quacks like a hash...

Comment (NobodysNightmare) [reply lib_static/open_project/configuration/helpers.rb]:
```
[2] pry(main)> h = HashWithIndifferentAccess.new
=> {}
[3] pry(main)> h[:foo] = 42
=> 42
[4] pry(main)> h
=> {"foo" => 42}
[5] pry(main)> h.transform_keys(&:to_sym)
=> {"foo" => 42}
[6] pry(main)> h.to_h
=> {"foo" => 42}
[7] pry(main)> h.symbolize_keys
=> {foo: 42}
[8] pry(main)> h.symbolize_keys.class
=> Hash
```

Comment (cbliard) [reply lib_static/open_project/configuration/helpers.rb]: You were faster than me. I was about to post the same kind of things.
Thanks for looking into it 👍

File: spec/workers/attachments/extract_fulltext_job_spec.rb (hunk 2 above)
Comment (cbliard) [reply spec/workers/attachments/extract_fulltext_job_spec.rb]: yes, it's all about test implementation. I wanted to reuse the shared example, but it may be better without using it. I'll try rewording it.

Comment (NobodysNightmare) [reply spec/workers/attachments/extract_fulltext_job_spec.rb]: Let me throw this idea into the ring:

```suggestion
    context "when ignoring the raised exception" do
```

Comment (cbliard) [reply spec/workers/attachments/extract_fulltext_job_spec.rb]: Is it better now?

File: spec/workers/attachments/extract_fulltext_job_spec.rb [hunk 3]
```diff
@@ -91,39 +102,47 @@
...
-    it "logs the error" do
-      extracted_attributes
+    it "raises the exception to update Job's internal state" do
+      expect { perform_enqueued_jobs }.to raise_error(exception_message)
+    end
+
+    it "logs the exception" do
+      perform_enqueued_jobs
+    rescue StandardError
       expect(logger).to have_received(:error).with(/boom-internal-error/)
```
Comment (NobodysNightmare) [inline spec/workers/attachments/extract_fulltext_job_spec.rb:136]: Note: This test (and the test below) will always pass if there is no exception raised, regardless of an error message being logged.

It's not too bad, because we validate the raising of an exception in another test case, but that's why I suggested the `perform_enqueued_jobs rescue nil` syntax before.

File: spec/workers/attachments/extract_fulltext_job_spec.rb (hunk 2 above)
Comment (NobodysNightmare) [reply spec/workers/attachments/extract_fulltext_job_spec.rb]: Yes, much clearer IMO.

**Review by NobodysNightmare — APPROVED** [review]

File: spec/workers/attachments/extract_fulltext_job_spec.rb (hunk 3 above)
Comment (cbliard) [reply spec/workers/attachments/extract_fulltext_job_spec.rb:136]: You're right. I've fixed it again and added a rubocop config file for `spec/` dir to ignore [Style/RescueModifier](https://docs.rubocop.org/rubocop/cops_style.html#stylerescuemodifier) (the `xxx rescue nil` thing).
